lib/spack/llnl/util/lang.py

```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
                 yield re.sub('.py$', '', name)
 
 
-def key_ordering(cls):
-    """Decorates a class with extra methods that implement rich comparison
-       operations and ``__hash__``.  The decorator assumes that the class
-       implements a function called ``_cmp_key()``.  The rich comparison
-       operations will compare objects using this key, and the ``__hash__``
-       function will return the hash of this key.
+#: sentinel for testing that iterators are done
+done = object()
+
+
+def tuplify(seq):
+    """Helper for lazy_lexicographic_ordering()."""
+    return tuple(tuplify(x) if callable(x) else x for x in seq())
+
+
+def doublewrap(f):
```

Comment:
```suggestion
def coerce_decorator_arguments(f):
```
along with @alalazo's suggestion!